airflow/dag_processing/collection.py

```diff
@@ -158,75 +140,101 @@ def calculate(cls, dags: dict[str, DAG], *, session: Session) -> Self:
+    orm_tags = {t.name: t for t in dm.tags}
+    for name, orm_tag in orm_tags.items():
+        if name not in tag_names:
+            session.delete(orm_tag)
+    dm.tags.extend(DagTag(name=name, dag_id=dm.dag_id) for name in tag_names if name not in orm_tags)
+
+
+def _update_dag_owner_links(dag_owner_links: dict[str, str], dm: DagModel, *, session: Session) -> None:
+    orm_dag_owner_attributes = {obj.owner: obj for obj in dm.dag_owner_links}
+    for owner, obj in orm_dag_owner_attributes.items():
+        try:
+            link = dag_owner_links[owner]
+        except KeyError:
+            session.delete(obj)
         else:
-            dm.calculate_dagrun_date_fields(dag, last_automated_data_interval)
-
-        if not dag.timetable.dataset_condition:
-            dm.schedule_dataset_references = []
-            dm.schedule_dataset_alias_references = []
-        # FIXME: STORE NEW REFERENCES.
-
-        dag_tags = set(dag.tags or ())
-        for orm_tag in (dm_tags := list(dm.tags or [])):
-            if orm_tag.name not in dag_tags:
-                session.delete(orm_tag)
-                dm.tags.remove(orm_tag)
-        orm_tag_names = {t.name for t in dm_tags}
-        for dag_tag in dag_tags:
-            if dag_tag not in orm_tag_names:
-                dag_tag_orm = DagTag(name=dag_tag, dag_id=dag.dag_id)
-                dm.tags.append(dag_tag_orm)
-                session.add(dag_tag_orm)
-
-        dm_links = dm.dag_owner_links or []
-        for dm_link in dm_links:
-            if dm_link not in dag.owner_links:
-                session.delete(dm_link)
-        for owner_name, owner_link in dag.owner_links.items():
-            dag_owner_orm = DagOwnerAttributes(dag_id=dag.dag_id, owner=owner_name, link=owner_link)
-            session.add(dag_owner_orm)
+            if obj.link != link:
+                obj.link = link
+    dm.dag_owner_links.extend(
+        DagOwnerAttributes(dag_id=dm.dag_id, owner=owner, link=link)
+        for owner, link in dag_owner_links.items()
+        if owner not in orm_dag_owner_attributes
+    )
+
+
+class DagModelOperation(NamedTuple):
+    """Collect DAG objects and perform database operations for them."""
+
+    dags: dict[str, DAG]
+
+    def add_dags(self, *, session: Session) -> dict[str, DagModel]:
+        orm_dags = _find_orm_dags(self.dags, session=session)
+        orm_dags.update(
+            (model.dag_id, model)
+            for model in _create_orm_dags(
+                (dag for dag_id, dag in self.dags.items() if dag_id not in orm_dags),
+                session=session,
+            )
+        )
+        return orm_dags
+
+    def update_dags(
+        self,
+        orm_dags: dict[str, DagModel],
+        *,
+        processor_subdir: str | None = None,
+        session: Session,
+    ) -> None:
+        run_info = _RunInfo.calculate(self.dags, session=session)
+
+        for dag_id, dm in sorted(orm_dags.items()):
```

Comment:
May I know why we want to sort it here?

Reply:
I don’t know either, but it’s been sorted since… forever? Probably not worthwhile to potentially break something.

Reply:
sounds reasonable